Add JdkInstaller

joycetoh8 opened the pull request:
If JDK 8 is not already installed, JdkInstaller will download
the binary and source code at given path.

PEConn commented on packages/core/src/lib/jdk/JdkInstaller.ts:
```diff
@@ -0,0 +1,96 @@
+  private unzipFunction: (srcPath: string, dstPath: string, deleteWhenDone: boolean)
+    => Promise<void>;
+  private joinPath: (...paths: string[]) => string;
+  /**
```
nit: newline before the comment.

PEConn commented on packages/core/src/lib/jdk/JdkInstaller.ts:
```diff
@@ -0,0 +1,96 @@
+    => Promise<void>;
+  private joinPath: (...paths: string[]) => string;
+  /**
+   * Constructs a new instance of JdkInstaller
```
nit: full stops at the end of sentences (here, line 46, line 77).

PEConn commented on packages/core/src/lib/jdk/JdkInstaller.ts:
```diff
@@ -0,0 +1,96 @@
+      }
+      default:
+        this.downloadFile = '';
+        console.log('Platform not found. Cannot download appropriate JDK.');
```
Would it not be better to throw the error at this point instead of in `install`? We know that the `JdkInstaller` is going to be useless here, I don't see the point in letting it exist for a while in an invalid state.

PEConn commented on packages/core/src/lib/jdk/JdkInstaller.ts:
```diff
@@ -0,0 +1,96 @@
+const JDK_SRC_ZIP = `jdk${JDK_VERSION}.zip`;
+
+/**
+ * Checks whether JDK 8 is installed. If not installed,
```
This comment is wrong isn't it? I can't see where this class checks whether JDK 8 is installed.

PEConn commented on packages/core/src/lib/jdk/JdkInstaller.ts:
```diff
@@ -0,0 +1,96 @@
+/**
+ * Checks whether JDK 8 is installed. If not installed,
+ * download the binary and source code and decompress at
+ * path given by user.
```
Do you mind adding a comment saying why we need to download the JDK source code? Otherwise I could see some well intentioned person coming along and removing it since it's unused.

joycetoh8 replied:
Good point! I've added a line about it

PEConn commented on packages/core/src/lib/jdk/JdkInstaller.ts:
```diff
@@ -0,0 +1,96 @@
+
+/**
+ * Checks whether JDK 8 is installed. If not installed,
+ * download the binary and source code and decompress at
```
nit: your line wrapping is a bit weird here.

andreban commented:
I'm wondering if we should assume the "default" flow as installing the JDK to a default place. This would make the user flow easier. Something like:

> Do you want Bubblewrap to download / install a JDK now? (Answer No to re-use an existing JDK installation). (Y/n).

If the user answers yes (default), we can just install that into `$HOME/.bubblewrap/jdk` to make it simpler. If the user answers No, we ask them for the JDK Path.

joycetoh8 commented:
@PEConn thanks for the review! I think I've addressed your comments now.

@andreban I also considered that but ultimately went with letting the user choose the location to have more choice. But I think you're right. Making it install in a default location is simpler for the user. So I've changed it to do this instead.

andreban commented on packages/core/src/lib/jdk/JdkInstaller.ts:
```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2019 Google Inc. All Rights Reserved.
```
nit: wrong year for the copyright header.

```suggestion
 * Copyright 2020 Google Inc. All Rights Reserved.
```

andreban reviewed (approved):
LGTM, with a nit on the date on the header of JdkInstaller.

Thank you!
